tests/integration/dex.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from os.path import join
+from pathlib import Path
+from time import sleep
+from typing import List, Optional
+
+import requests
+from lightkube import Client, codecs
+from lightkube.core.exceptions import ApiError, ObjectDeleted
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Pod, Service
+from requests.exceptions import RequestException
+
+logger = logging.getLogger(__name__)
+
+
+DEX_MANIFESTS = Path(__file__).parent.parent.parent / "manifests" / "dex.yaml"
+
+
+def get_dex_manifest(
+    client_id: Optional[str] = None,
+    client_secret: Optional[str] = None,
+    redirect_uri: Optional[str] = None,
+    issuer_url: Optional[str] = None,
+) -> List[codecs.AnyResource]:
+    with open(DEX_MANIFESTS, "r") as file:
+        return codecs.load_all_yaml(
+            file,
+            context={
+                "client_id": client_id,
+                "client_secret": client_secret,
+                "redirect_uri": redirect_uri,
+                "issuer_url": issuer_url,
+            },
+        )
+
+
+def dex_is_ready(client: Client, issuer_url: Optional[str] = None) -> None:
+    for pod in client.list(Pod, namespace="dex", labels={"app": "dex"}):
+        # Some pods may be deleted, if we are restarting
+        try:
+            client.wait(
+                Pod, pod.metadata.name, for_conditions=["Ready", "Deleted"], namespace="dex"
+            )
+        except ObjectDeleted:
+            pass
+    client.wait(Deployment, "dex", namespace="dex", for_conditions=["Available"])
+    if not issuer_url:
+        issuer_url = get_dex_service_ip(client)
+
+    resp = requests.get(join(issuer_url, ".well-known/openid-configuration"))
+    if resp.status_code != 200:
+        raise RuntimeError("Failed to deploy dex")
+
+
+def apply_dex_resources(
```

Comment:
I guess this function assumes more responsibilities than it should when simply looking at the function name. It includes the logic for applying k8s manifests, restarting k8s resources or not, and waiting for the resources to be ready or not. It's better to divide it into three individual functions I guess because the user would expect this function just simply `apply` the manifests.